drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2019-2020 Peter Bigot Consulting, LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT regulator_gpio
+
+#include <drivers/regulator.h>
+#include <drivers/gpio.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(regulator_gpio, CONFIG_REGULATOR_LOG_LEVEL);
+
+#define OPTION_BOOT_ALWAYS_ON_POS 0
+#define OPTION_BOOT_ALWAYS_ON BIT(OPTION_BOOT_ALWAYS_ON_POS)
+#define OPTION_BOOT_ON_POS 0
+#define OPTION_BOOT_ON BIT(OPTION_BOOT_ON_POS)
+
+struct driver_config {
+	const char *gpio_name;
+	uint32_t delay_us;
+	gpio_pin_t gpio_pin;
+	gpio_dt_flags_t gpio_flags;
+	uint8_t options;
+};
+
+struct driver_data {
+	struct device *dev;
+	struct device *gpio;
+	struct onoff_manager mgr;
+};
+
+/* Declaration here, definition after members defined. */
+static const struct regulator_driver_api api;
+
+static void start(struct onoff_manager *mgr,
+		  onoff_notify_fn notify)
+{
+	const struct driver_data *data =
+		CONTAINER_OF(mgr, struct driver_data, mgr);
+	const struct driver_config *cfg = data->dev->config_info;
+	int rc = gpio_pin_set(data->gpio, cfg->gpio_pin, true);
+
+	if (rc == -EWOULDBLOCK) {
+		/* TODO: delegate completion to a work queue. */
+	} else if ((rc >= 0) && (cfg->delay_us > 0)) {
+		if (k_is_in_isr()
+		    || k_is_pre_kernel()
+		    || (k_us_to_ticks_floor32(cfg->delay_us) == 0)) {
+			k_busy_wait(cfg->delay_us);
+		} else {
+			/* NB: This may sleep longer than desired, as
+			 * resolution is in multiple of ticks.
+			 */
+			k_usleep(cfg->delay_us);
+		}
+	}
+
+	notify(mgr, rc);
+}
+
+static void stop(struct onoff_manager *mgr,
+		 onoff_notify_fn notify)
+{
+	const struct driver_data *data =
+		CONTAINER_OF(mgr, struct driver_data, mgr);
+	const struct driver_config *cfg = data->dev->config_info;
+	int rc = gpio_pin_set(data->gpio, cfg->gpio_pin, false);
+
+	/* TODO: detect -EWOULDBLOCK and delegate completion to a work
+	 * queue.
+	 */
+	if (rc == -EWOULDBLOCK) {
+		/* TODO: delegate completion to a work queue. */
+	}
+	notify(mgr, rc);
+}
+
+static int enable(struct device *dev, struct onoff_client *cli)
+{
+	struct driver_data *data = dev->driver_data;
+	int rc = -EINVAL;
+
+	if (dev->driver_api == &api) {
+		rc = onoff_request(&data->mgr, cli);
+	}
+
+	return rc;
+}
+
+static int disable(struct device *dev)
```

Comment:
With "always on" bit set, should the driver take care of returning error? Or can it be handled in `disable()` instead?

Reply:
For now I'm going to do it in the driver; if we get more drivers then it could be lifted to the wrapper.